forge-ai/src/main/java/forge/ai/AiController.java

```diff
@@ -1574,79 +1605,99 @@ private SpellAbility chooseSpellAbilityToPlayFromList(final List<SpellAbility> a
+                                aiPlayDecision = AiPlayDecision.CantAfford;
+                            } else {
+                                aiPlayDecision = AiPlayDecision.WillPlay;
+                            }
+                        }
+                    } else if (sa.getHostCard().hasKeyword(Keyword.CASCADE)) {
+                        if (isLifeInDanger) { //needs more tune up for certain conditions
+                            aiPlayDecision = player.getCreaturesInPlay().size() >= 4 ? AiPlayDecision.CantPlaySa : AiPlayDecision.WillPlay;
+                        } else if (CardLists.filter(player.getZone(ZoneType.Graveyard).getCards(), CardPredicates.Presets.CREATURES).size() > 4) {
+                            if (player.getCreaturesInPlay().size() >= 4) // it's good minimum
+                                return 0;
+                            else if (!sa.getHostCard().isPermanent() && sa.canCastTiming(player) && ComputerUtilCost.canPayCost(sa, player, sa.isTrigger()))
+                                aiPlayDecision = AiPlayDecision.WillPlay;// needs tuneup for bad matchups like reanimator and other things to check on opponent graveyard
                         } else {
-                            aiPlayDecision = AiPlayDecision.WillPlay;
+                            return 0;
                         }
                     }
-                } else if (sa.getHostCard().hasKeyword(Keyword.CASCADE)) {
-                    if (isLifeInDanger) { //needs more tune up for certain conditions
-                        aiPlayDecision = player.getCreaturesInPlay().size() >= 4 ? AiPlayDecision.CantPlaySa : AiPlayDecision.WillPlay;
-                    } else if (CardLists.filter(player.getZone(ZoneType.Graveyard).getCards(), CardPredicates.Presets.CREATURES).size() > 4) {
-                        if (player.getCreaturesInPlay().size() >= 4) // it's good minimum
-                            continue;
-                        else if (!sa.getHostCard().isPermanent() && sa.canCastTiming(player) && ComputerUtilCost.canPayCost(sa, player, sa.isTrigger()))
-                            aiPlayDecision = AiPlayDecision.WillPlay;// needs tuneup for bad matchups like reanimator and other things to check on opponent graveyard
-                    } else {
-                        continue;
-                    }
                 }
-            }
 
-            sa.setActivatingPlayer(player, true);
-            SpellAbility root = sa.getRootAbility();
+                // dumb check needed? don't really know why the AI tapped the land for nothing at EOT of his opponent.
+                // without this, you can see the weird tapping of single land -> mana ability. with this, it doesn't confuse human player
+                if (sa.isManaAbility() && game.getPhaseHandler() != null && game.getPhaseHandler().is(PhaseType.END_OF_TURN) && !game.getPhaseHandler().isPlayerTurn(player))
```

Comment:
are you saying it only happens on this branch?
then some logic got messed up somehow and should be fixed instead...
otherwise I can try and fix it on master

Reply:
I was thinking on the sorting of SA. On this branch I moved the sort after the SA is gathered with timeout since the order will not be the same and needs to be sorted again. Gonna try tomorrow again.

Reply:
I don't want to argue on what you believe but the point of this change is to have a timeout regardless if the boardstate is complex or not. The current problem is where the number to process those objects using sequential approach will eventually be slower than processing them concurrently and anything that's within the timeout will have results than waiting forever to wait for the game to proceed. Also the threaded computation are only specifically selected on the most expensive computation so it will still have results of it timeout or not.  This was intended to make the game move on those state where it's hard to process  all. I don't know how long you play the game but it's tiresome to wait for the AI if it's not producing results without timeout. All this change was to have a better play experience than having a perfect AI logic that will stall over time if the boardstate becomes to much to process.

Reply:
yes I understand all that and I'm not against it

frankly often when I debugged some slowdown it was at least partially caused by very few cards/API which run a badly optimized path (often custom written AILogic$)

but anyway my point was why another structural change on top of this?
why force check all API first when there's a good chance you could find a playable candidate way before the last element?
is it because you have no control over the order the Futures get calculated so you can't just return with one of the first matches?

Reply:
Because you don't know what will be finish first specially if the spellabilities is big. Example you have 10 creatures that have deals damage to target creature. And the opponent have a lot of possible target. If the first one didn't find good target, it will take the next one again and will have the same result as the first one leaving the computation to process all of them sequentially to just not find anything feasible. If you process them concurrently, you will have faster results. The cons of this approach is if those abilities are unique, you need to sort them after. unlike the previous implementation that is already sorted (if verify transitivity didn't occur) and if the logic passes, it will be returned, the cons of that approach is if the top sa will fail the logic, it will process the next one and wait if it passes or not and so on.